### PR TITLE
feat(5226): more verbose line error breakdown returned on write request

### PR DESCRIPTION
**Pull request by wiedld**
Part of #5226 

The objective of #5226 is to provide the ability to enable large writes to partially succeed upon request. Specifically:
* set env var => writes succeed for any non-erroring line (a.k.a. partial writes, closer to v1/v2 compatibility).
* use the default setting => all-or-none write success (a.k.a. current iox behavior)

As a precursor to this work, this PR is enabling a more detailed breakdown (bounded at 100 line errors) of line protocol failures. This per-line error breakdown will occur regardless of which mode we are in.

The followup PR will introduce the ability to do all-or-nothing versus partial. This PR is only about returning itemized line errors.

---

* test(5226): demonstrate current API contract for lp writes (c786e4261)

* test(5226): include message assertion within write end-to-end tests (ea64c864b)

* feat(5226): enable payload body to be returned on HttpApiError for writes (3d3769f51)

* feat(5226): update mutable_lp LinesConverter to return multiple line errors (34654339e)
     ```
      * This required have a new LineError type, such that the mutable_batch_lp::Error could have a PerLine(vec of LineErrors).
      * Updates the contract to have LinesConverter::write_lp() continue processing lines, and logging errors, up until max.
      * Writes are still atomic. Will reject a write based upon an error; it simply collects multiple errors.
     ```

* test(5226): update test_write_handler and test_http_handler tests to handle the new format of multiple LineErrors being returned (6ef7d3549)

* feat(5226): bring API into closer alignment with v1/v2, with a 'line' in payload (a0ff74de5)

### CODE REVIEW

File: router/src/server/http.rs [hunk 1]
```diff
@@ -346,11 +379,11 @@ where
         converter.set_timestamp_base(write_info.precision.timestamp_base());
         let (batches, stats) = match converter.write_lp(body).and_then(|_| converter.finish()) {
             Ok(v) => v,
-            Err(mutable_batch_lp::Error::EmptyPayload) => {
+            Err(e) if matches!(e, mutable_batch_lp::Error::EmptyPayload) => {
                 debug!("nothing to write");
                 return Ok(());
             }
-            Err(e) => return Err(Error::ParseLineProtocol(e)),
+            Err(line_errors) => return Err(Error::ParseLineProtocol(line_errors)),
```
Comment (wiedld) [inline router/src/server/http.rs:386]: This line is where the all-or-nothing occurs for the writes. It returns an error, such we never hit the `dml_handler.write()`.

File: router/src/server/http.rs [hunk 2]
```diff
@@ -515,7 +548,7 @@ mod tests {
             body = $body:expr,                              // Request body content
             dml_write_handler = $dml_write_handler:expr,    // DML write handler response (if called)
             dml_delete_handler = $dml_delete_handler:expr,  // DML delete handler response (if called)
-            want_result = $want_result:pat,                 // Expected handler return value (as pattern)
+            want_result = [$($want_result:tt )+],           // Expected handler return value (as pattern)
```
Comment (wiedld) [inline router/src/server/http.rs:551]: To add more flexibility in what can be used later (in `assert_matches!(got, $want_result)`), this was changed to a tt muncher. Since we now have two tt munchers in a row, the token recognition needed a bit more help to delineate the boundaries. Hence the added `[]`.

File: router/src/server/http.rs [hunk 3]
```diff
@@ -1411,8 +1472,35 @@ mod tests {
...
+                        },
+                        line: 44,
+                    },
+                ]
+            }),
+            "failed to parse line protocol: \
+            errors encountered on 3 lines:\
+            \nerror parsing line 42 (1-based): No fields were provided\
+            \ntimestamp overflows i64 on line 43 (1-based)\
+            \nerror writing line 44 (1-based): the field 'bananas' is specified more than once with conflicting types",
```
Comment (wiedld) [inline router/src/server/http.rs:1503]: Note that the stdout will still show the multiple lines of errors.

File: influxdb_iox/tests/end_to_end_cases/router.rs [hunk 4]
```diff
@@ -47,6 +49,60 @@ pub async fn test_json_errors() {
...
+            },
+            Step::WriteLineProtocolExpectingError {
+                line_protocol: "table_atomic,tag1=B,tag2=good val=44i 123458\n\
+                     ,tag1=B,tag2=bad val=45i 123459\n\
+                     table_atomic,tag1=B,tag2=good val=46i 123460"
+                    .into(),
+                expected_error_code: StatusCode::BAD_REQUEST,
+                expected_error_message: "failed to parse line protocol: errors encountered on 1 line".to_string(),
+                expected_error_body: HashMap::<String, String>::from([
+                    ("line_errors".into(), "[{\"err\":\"error parsing line 2 (1-based): Invalid measurement was provided\",\"index\":2}]".into())
```
Comment (wiedld) [inline influxdb_iox/tests/end_to_end_cases/router.rs:86]: Note the http response has the error message versus error body split.
The test is identical to the stdout, just split into parts.

**Comment by philjb** [comment]
explored the build - looks sweet

```
{
  "code": "invalid",
  "line_errors": [
    {
      "err": "error parsing line 1 (1-based): Tag Set Malformed",
      "index": 1
    },
    {
      "err": "error parsing line 3 (1-based): A generic parsing error occurred: TakeWhile1",
      "index": 3
    },
    {
      "err": "error parsing line 4 (1-based): A generic parsing error occurred: TakeWhile1",
      "index": 4
    },
    {
      "err": "error parsing line 5 (1-based): No fields were provided",
      "index": 5
    }
  ],
  "message": "failed to parse line protocol: errors encountered on 4 lines"
}
```

 (I formatted it from the original error message)

**Comment by wiedld** [comment]
Per [clarified requirements](https://github.com/influxdata/influxdb_iox/issues/5226#issuecomment-1716158763) (TY!), we have tried to balance the needs for a more verbose (multi-line) errors -- while maintaining v1/v2 compatibility.

As such, the multi-line output is within the `message` returned. Specifically is:
```
{
    "code": "invalid",
    "line": 4,
    "message": "failed to parse line protocol: errors encountered on 3 lines:\nerror parsing line 4 (1-based): Invalid measurement was provided\nerror parsing line 7 (1-based): Unable to parse timestamp value '15918943100000000000000000'\nerror parsing line 13 (1-based): Unable to parse timestamp value '1591894310000000000000'"
}
```

Where the message (formatted newlines) is:
```
failed to parse line protocol: errors encountered on 3 lines:
error parsing line 4 (1-based): Invalid measurement was provided
error parsing line 7 (1-based): Unable to parse timestamp value '15918943100000000000000000'
error parsing line 13 (1-based): Unable to parse timestamp value '1591894310000000000000'
```

File: mutable_batch_lp/src/lib.rs [hunk 5]
```diff
@@ -26,24 +26,50 @@ use mutable_batch::writer::Writer;
...
 #[allow(missing_docs)]
 pub enum Error {
+    #[snafu(display(
+        "errors encountered on {}{} line{}:\n{}",
+        std::cmp::min(MAXIMUM_RETURNED_ERRORS, lines.len()),
+        if lines.len() > MAXIMUM_RETURNED_ERRORS { "+" } else { "" },
+        if lines.len() > 1 { "s" } else { "" },
+        lines.iter().map(|line| line.to_string()).collect::<Vec<String>>().join("\n")
+    ))]
+    PerLine { lines: Vec<LineError> },
```
Comment (domodwyer) [inline mutable_batch_lp/src/lib.rs:42]: This is... not pretty! Can we generate the string in code once, and add that to the error instead of adding all the errors and generating the string each time it needs rendering?

File: mutable_batch_lp/src/lib.rs [hunk 6]
```diff
@@ -26,24 +26,50 @@ use mutable_batch::writer::Writer;
...
 
-    #[snafu(display("empty write payload"))]
-    EmptyPayload,
+    #[snafu(display("timestamp overflows i64 on line {} (1-based)", line))]
+    TimestampOverflow { line: usize },
+}
 
-    #[snafu(display("timestamp overflows i64"))]
-    TimestampOverflow,
+macro_rules! break_or_continue {
```
Comment (domodwyer) [inline mutable_batch_lp/src/lib.rs:65]: If you're reaching for a macro, it's usually a sign that something is structured incorrectly. I know it makes sense to you, but reading the call sites of `break_or_continue(x > 2)` is not particularly clear about what happens if true or false for others, and the logic is hard to follow.

The `write_lp` block below with the loops, anonymous function/closure, macro invocations, nested conditionals and nested control flow could instead be expressed as an iterator, which would be both idiomatic, and more performant:

```rust
    let mut errors_so_far = 0;
    let errors = parse_lines(lines)
        .enumerate()
        .map(parse_line)
        .take_while(|res| {
            if res.is_err() {
                errors_so_far += 1;
                return errors_so_far < MAXIMUM_RETURNED_ERRORS;
            }

            true
        })
        .filter(Result::is_err)
        .collect::<Vec<_>>();
```

This will give you an array of errors (up to `MAXIMUM_RETURNED_ERRORS`) or an empty array for no errors, evaluating only as much data as needed (returns early). 

It's also _way easier_ to read and reason about, because this separates the logic of parsing the actual lines, from the logic of error handling/accumulation - separation of concerns!

Ideally you further compose up small functions to do the actual parsing:

```rust
fn parse_line(idx: usize, maybe_line: Result<ParsedLine<'_>, Error>) -> ... {
    maybe_line
        .and_then(|line| rebase_timestamp(idx, line, timestamp_base))
        .and_then(|...| add_line_to_batch(...))
}
```

So the overall structure is a composition of small, singly responsible blocks of code that do _one thing_ - it's easy to read and understand, and has a low risk of bugs

File: mutable_batch_lp/src/lib.rs [hunk 7]
```diff
@@ -107,13 +133,35 @@ impl LinesConverter {
...
     pub fn write_lp(&mut self, lines: &str) -> Result<()> {
+        // TODO: next PR will have this be configurable per env var.
+        let abort_on_line_error = true;
+
+        let mut errors: Vec<LineError> = Vec::with_capacity(MAXIMUM_RETURNED_ERRORS);
+        let mut add_error = |e: LineError| {
+            if errors.len() < MAXIMUM_RETURNED_ERRORS {
+                errors.push(e);
+            }
+            abort_on_line_error && errors.len() >= MAXIMUM_RETURNED_ERRORS
```
Comment (domodwyer) [inline mutable_batch_lp/src/lib.rs:144]: I'm not sure this is correct - perhaps an `if` would make it clearer?

This currently says:

"if abort_on_line_error == true AND more errors than the max allowed"

I think we're expecting one OR the other to stop parsing? Otherwise won't it evaluate all errors when `abort_on_line_error == false` (but not add them to the set) instead of returning early? 

I might be wrong, because it's not very clear going through the macro and loops :fire:

File: mutable_batch_lp/src/lib.rs [hunk 8]
```diff
@@ -107,13 +133,35 @@ impl LinesConverter {
     ///     [`mutable_batch::writer::Error::TypeMismatch`]
     ///
     pub fn write_lp(&mut self, lines: &str) -> Result<()> {
+        // TODO: next PR will have this be configurable per env var.
+        let abort_on_line_error = true;
+
+        let mut errors: Vec<LineError> = Vec::with_capacity(MAXIMUM_RETURNED_ERRORS);
```
Comment (domodwyer) [inline mutable_batch_lp/src/lib.rs:139]: I don't think we should be allocating the vector here - optimise for the happy path - there's no need to allocate 100 errors up front when 99% of our requests will never use it. 

Optimise for the fast, common path 👍

File: ioxd_common/src/http/error.rs [hunk 9]
```diff
@@ -109,14 +123,23 @@ impl HttpApiError {
...
             code: code.into(),
             msg: msg.into(),
+            body: Default::default(),
         }
     }
 
+    /// Add body to error.
+    pub fn with_body(self, raw: HashMap<String, impl Into<serde_json::Value>>) -> Self {
+        let body: HashMap<String, serde_json::Value> =
+            raw.into_iter().map(|(k, v)| (k, v.into())).collect();
```
Comment (domodwyer) [inline ioxd_common/src/http/error.rs:133]: This will force us to iterate over all the things we give to the body because of the `Into`, when we probably already had to iterate to generate the hashmap anyway - we should just pass in the right thing, and avoid loping over the same data multiple times :+1:

File: ioxd_common/src/http/error.rs [hunk 10]
```diff
@@ -109,14 +123,23 @@ impl HttpApiError {
...
+
     /// Generate response body for this error.
     fn body(&self) -> Body {
-        let json = serde_json::json!({
-            "code": self.code.as_text().to_string(),
-            "message": self.msg.clone(),
+        let json = serde_json::json!(HttpApiErrorJson {
+            code: self.code.as_text().to_string(),
+            message: self.msg.clone(),
+            body: self.body.clone(),
```
Comment (domodwyer) [inline ioxd_common/src/http/error.rs:142]: The `HttpApiErrorJson` instance doesn't escape this function, nor does it ever get used anywhere else - we should borrow the body instead of cloning all the error strings again.

File: router/src/server/http.rs [hunk 11]
```diff
@@ -131,6 +131,27 @@ impl Error {
...
+        match self {
+            Self::ParseLineProtocol(mutable_batch_lp::Error::PerLine { lines }) => {
+                let mut values = StdHashMap::<String, serde_json::Value>::default();
+                let line = match lines.get(0) {
+                    Some(LineError::LineProtocol { source: _, line })
+                    | Some(LineError::TimestampOverflow { line })
+                    | Some(LineError::Write { source: _, line }) => *line,
+                    None => unreachable!("PerLine error must have at least one line"),
+                };
+                values.insert("line".into(), serde_json::Value::Number(line.into()));
```
Comment (domodwyer) [inline router/src/server/http.rs:149]: Why do we return a hashmap here if there's only ever one entry, and it's always got a static key? 

Shouldn't it be `-> Option<usize>` instead? 

It also probably shouldn't be coupled with the encoding type (`json::Value`) because it becomes impossible to reuse outside of a HTTP JSON API (and the error shouldn't care about _how it is represented_ - it just says what went wrong - something else is responsible for turning that into a user-facing representation).

File: router/Cargo.toml [hunk 12]
```diff
@@ -29,6 +29,7 @@ object_store = { workspace = true }
 observability_deps = { path = "../observability_deps" }
 parking_lot = "0.12"
 serde = "1.0"
+serde_json = "1.0.105"
```
Comment (domodwyer) [inline router/Cargo.toml:32]: The router crate isn't responsible for the API / encoding / how it is used - it's role is to process data and return errors (if any) and the HTTP layer maps that into an error - that's why there's no JSON encoding in here today, for a system that already returns JSON.

File: router/src/server/http.rs [hunk 13]
```diff
@@ -346,11 +367,11 @@ where
         converter.set_timestamp_base(write_info.precision.timestamp_base());
         let (batches, stats) = match converter.write_lp(body).and_then(|_| converter.finish()) {
             Ok(v) => v,
-            Err(mutable_batch_lp::Error::EmptyPayload) => {
+            Err(e) if matches!(e, mutable_batch_lp::Error::EmptyPayload) => {
```
Comment (domodwyer) [inline router/src/server/http.rs:370]:
```suggestion
            Err(mutable_batch_lp::Error::EmptyPayload) => {
```

File: router/src/server/http.rs (hunk 11 above)
Comment (wiedld) [reply router/src/server/http.rs:149]: Changes made:
* The router is no longer dependent on serde_json, and instead have the conversion point (to the generic HashMap of serde_json::Values) occur within the ioxd_router.
* The ioxd_common remains generalized and takes any HashMap of serde_json::Values, to be flattened into the payload body.

File: mutable_batch_lp/src/lib.rs (hunk 6 above)
Comment (wiedld) [reply mutable_batch_lp/src/lib.rs:65]: That's much nicer. Thank you.

File: mutable_batch_lp/src/lib.rs (hunk 7 above)
Comment (wiedld) [reply mutable_batch_lp/src/lib.rs:144]: Already removed with refactoring.

File: router/src/server/http.rs (hunk 13 above)
Comment (wiedld) [reply router/src/server/http.rs]: done.

File: mutable_batch_lp/src/lib.rs (hunk 5 above)
Comment (wiedld) [reply mutable_batch_lp/src/lib.rs:42]: I'm confused on what you are looking for here. Shouldn't display only be called once per request? (Unless there is a logging per request.)

Wouldn't storing as strings get us away from the well typed errors? Or if we store both the well-typed errors and a string, would this not increase allocation?

Comment (wiedld) [reply mutable_batch_lp/src/lib.rs]: I made the same string conversion be cheaper (without the vector allocation). See the latest commit.

Comment (domodwyer) [reply mutable_batch_lp/src/lib.rs]: My main objection is the very dense logic stuffed into the macro invocation - it's hard to read, and I'm not really sure why `lines` would ever be more than `MAXIMUM_RETURNED_ERRORS` in the first place, because we want to gather, at most, `MAXIMUM_RETURNED_ERRORS` so the logic seems unnecessary? If it's just to pluralise the string, just wang in `error(s)` and save the CPU/complexity :shrug:

**Review by domodwyer — COMMENTED** [review]
I think your life will be much easier if you do away with the `serde_json::Value` passing - it's forcing you to do things completely differently from how the API is currently structured.

File: ioxd_router/src/lib.rs [hunk 14]
```diff
@@ -217,7 +217,12 @@ impl std::error::Error for IoxHttpErrorAdaptor {}
 
 impl HttpApiErrorSource for IoxHttpErrorAdaptor {
     fn to_http_api_error(&self) -> HttpApiError {
-        HttpApiError::new(self.0.as_status_code(), self.to_string())
+        let mut body = std::collections::HashMap::<String, serde_json::Value>::new();
+        if let Some(line) = self.0.get_line() {
+            body.insert("line".into(), line.into());
+        };
+
+        HttpApiError::new(self.0.as_status_code(), self.to_string()).with_body(body)
```
[the remark this answers is not in the document]
Comment (wiedld) [reply ioxd_router/src/lib.rs:225]: ioxd_common error body accepts the hashmap, because our (overall) goal is to move the APIs to be v1/v2 compatible. Looking at other error payloads returned, I found additional keys (other than just `line` for an lp write error), including `err` in v1 & v2 query. 

As such, the flattening of a provided hashmap permits this ioxd_common  code to be generalized -- such that we can attached other keys in the returned error body.

[the remark this answers is not in the document]
Comment (wiedld) [reply ioxd_router/src/lib.rs:225]: Is the goal instead to have each additional key be a prop on `HttpApiError`? Such that all services only return types, but then ioxd_common has props compatible for each type?

File: ioxd_router/Cargo.toml [hunk 15]
```diff
@@ -20,6 +20,7 @@ metric = { path = "../metric" }
 mutable_batch = { path = "../mutable_batch" }
 object_store = { workspace = true }
 router = { path = "../router" }
+serde_json = "1.0.105"
```
[the remark this answers is not in the document]
Comment (wiedld) [reply ioxd_router/Cargo.toml:23]: removed json outside of ioxd_common (where it already exists).

File: mutable_batch_lp/src/lib.rs (hunk 5 above)
Comment (wiedld) [reply mutable_batch_lp/src/lib.rs]: Switched to using `error(s)`. 👍🏼

File: ioxd_common/src/http/error.rs [hunk 16]
```diff
@@ -109,18 +121,18 @@ impl HttpApiError {
...
     /// Generate response body for this error.
     fn body(&self) -> Body {
-        let json = serde_json::json!({
-            "code": self.code.as_text().to_string(),
-            "message": self.msg.clone(),
-        })
-        .to_string();
-
-        Body::from(json)
+        Body::from(serde_json::json!(self).to_string())
```
Comment (domodwyer) [inline ioxd_common/src/http/error.rs:135]: I believe this would normally use `serde_json::to_string()`:

```suggestion
        Body::from(serde_json::to_string(&self).expect("must serialise to json"))
```

File: ioxd_common/src/http/error.rs [hunk 17]
```diff
@@ -91,16 +92,27 @@ impl From<StatusCode> for HttpApiErrorCode {
...
     /// Machine-readable error code.
     code: HttpApiErrorCode,
 
     /// Human-readable message.
+    #[serde(rename = "message")]
     msg: String,
+
+    /// Optional error line (for line protocol errors).
+    #[serde(skip_serializing_if = "Option::is_none")]
+    line: Option<usize>,
```
Comment (domodwyer) [inline ioxd_common/src/http/error.rs:115]: In the future, instead of this growing to contain a field for every possible error field that could ever be returned in any API error response eror, this should take a `T` that is `T: Serializable`, with handler-specific payloads - this lets us have API handlers specify a struct of additional metadata that is returned for that call.

This will do for now though

File: router/src/server/http.rs [hunk 18]
```diff
@@ -131,6 +131,25 @@ impl Error {
...
     }
+
+    /// Return the line number of the first line that failed to parse,
+    /// if the error is a line protocol error.
+    ///
+    /// Line number is index 1 based.
+    pub fn get_parse_error_line_index(&self) -> Option<usize> {
+        match self {
+            Self::ParseLineProtocol(mutable_batch_lp::Error::PerLine { lines }) => {
+                let line = match lines.get(0) {
```
Comment (domodwyer) [inline router/src/server/http.rs:142]: You could `get(0).expect("message")` here and not have to wrap every match arm in a `Some` :+1:

**Review by domodwyer — APPROVED** [review]
Got there! Much cleaner - thanks for sticking with it 👍

File: ioxd_common/src/http/error.rs (hunk 17 above)
Comment (wiedld) [reply ioxd_common/src/http/error.rs:115]: Agreed. 
I actually tried t[hat approach](https://github.com/influxdata/influxdb_iox/pull/8690#discussion_r1329007627), with generics and default types, and it spread trait bounds into the server & service definitions etc. I wasn't sure if that was the right direction. 🤷🏼‍♀️
